shell/common/asar/archive_win.cc

```diff
@@ -0,0 +1,179 @@
+  }
+
+  base::FilePath exe_dir = base::FilePath(exe_path).DirName();
+  base::FilePath relative_path;
+  if (!exe_dir.AppendRelativePath(path_, &relative_path)) {
+    return absl::nullopt;
+  }
+
+  return relative_path;
+}
+
+void LoadIntegrityConfigCache() {
+  // Skip loading if cache is already loaded
+  if (s_win_integrity_payloads_cache->has_value()) {
+    return;
+  }
+
+  // Init cache
+  *s_win_integrity_payloads_cache = unordered_map<string, IntegrityPayload>();
+
+  // Load integrity config from exe resource
+  HMODULE module_handle = ::GetModuleHandle(NULL);
+
+  HRSRC resource = ::FindResource(module_handle, INTEGRITY_CHECK_RESOURCE_ITEM,
+                                  INTEGRITY_CHECK_RESOURCE_TYPE);
+  if (!resource) {
+    LOG(FATAL) << "FindResource failed. Last error: " << GetLastError();
+    return;
+  }
+
+  HGLOBAL rcData = ::LoadResource(module_handle, resource);
+  if (!rcData) {
+    LOG(FATAL) << "LoadResource failed. Last error: " << GetLastError();
+    return;
+  }
+
+  auto* res_data = static_cast<const char*>(::LockResource(rcData));
+  int res_size = SizeofResource(NULL, resource);
+
+  if (!res_data || !res_size) {
+    LOG(FATAL) << "Failed to integrity config from exe resource.";
+    return;
+  }
+
+  // Parse integrity config payload
+  auto integrity_config_payload = string(res_data, res_size);
+  absl::optional<base::Value> root = base::JSONReader::Read(
+      integrity_config_payload,
+      base::JSON_PARSE_CHROMIUM_EXTENSIONS | base::JSON_ALLOW_TRAILING_COMMAS);
+
+  if (!root.has_value()) {
+    LOG(FATAL) << "Invalid integrity config: NOT a valid JSON.";
+    return;
+  }
+
+  const base::Value::List* file_configs = root.value().GetIfList();
+  if (!file_configs) {
+    LOG(FATAL) << "Invalid integrity config: NOT a list.";
+    return;
+  }
+
+  // Parse each individual file integrity config
+  for (size_t i = 0; i < file_configs->size(); i++) {
+    // Skip invalid file configs
+    const Value::Dict* ele_dict = (*file_configs)[i].GetIfDict();
+    if (!ele_dict) {
+      LOG(WARNING) << "Skip config " << i << ": NOT a valid dict";
+      continue;
+    }
+
+    const string* file = ele_dict->FindString("file");
+    if (!file || file->empty()) {
+      LOG(WARNING) << "Skip config " << i << ": Invalid file";
+      continue;
+    }
+
+    const string* alg = ele_dict->FindString("alg");
+    if (!alg || base::ToLowerASCII(*alg) != "sha256") {
+      LOG(WARNING) << "Skip config " << i << ": Invalid alg";
+      continue;
+    }
+
+    const string* value = ele_dict->FindString("value");
+    if (!value || value->empty()) {
+      LOG(WARNING) << "Skip config " << i << ": Invalid hash value";
+      continue;
+    }
+
+    // Add valid file config into cache
+    IntegrityPayload header_integrity;
+    header_integrity.algorithm = HashAlgorithm::kSHA256;
+    header_integrity.hash = *value;
+
+    s_win_integrity_payloads_cache->value()[ToLowerASCII(*file)] =
+        header_integrity;
```

Comment:
I think this should be moved?

```suggestion
        std::move(header_integrity);
```

Reply:
Will test it

Reply:
Updated and tested